storm-client/src/jvm/org/apache/storm/utils/JCQueue.java

```diff
@@ -0,0 +1,457 @@
+        metrics.close();
+        return res;
+    }
+
+
+    /**
+     * Non blocking. Returns immediately if Q is empty. Returns number of elements consumed from Q
+     */
+    public int consume(JCQueue.Consumer consumer) {
+        return consume(consumer, continueRunning);
+    }
+
+    /**
+     * Non blocking. Returns immediately if Q is empty. Runs till Q is empty OR exitCond.keepRunning() return false.
+     * Returns number of elements consumed from Q
+     */
+    public int consume(JCQueue.Consumer consumer, ExitCondition exitCond) {
+        try {
+            return consumeImpl(consumer, exitCond);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public int size() { return recvQueue.size() + overflowQ.size(); }
+
+    /**
+     * Non blocking. Returns immediately if Q is empty. Returns number of elements consumed from Q
+     *  @param exitCond
+     */
+    private int consumeImpl(Consumer consumer, ExitCondition exitCond) throws InterruptedException {
+        int drainCount = 0;
+        while ( exitCond.keepRunning() ) {
+            Object tuple = recvQueue.poll();
+            if (tuple == null) {
+                break;
+            }
+            consumer.accept(tuple);
+            ++drainCount;
+        }
+
+        int overflowDrainCount = 0;
+        int limit = overflowQ.size();
+        while (exitCond.keepRunning()  &&  (overflowDrainCount < limit)) { // 2nd cond prevents staying stuck with consuming overflow
+            Object tuple = overflowQ.poll();
+            ++overflowDrainCount;
+            consumer.accept(tuple);
+        }
+        int total = drainCount + overflowDrainCount;
+        if (total > 0) {
+            consumer.flush();
+        } else {
+            emptyMeter.record();
+        }
+        return total;
+    }
+
+    // Non Blocking. returns true/false indicating success/failure. Fails if full.
+    private boolean tryPublishInternal(Object obj) {
+        if (recvQueue.offer(obj)) {
+            metrics.notifyArrivals(1);
+            return true;
+        }
+        return false;
+    }
+
+    // Non Blocking. returns count of how many inserts succeeded
+    private int tryPublishInternal(ArrayList<Object> objs) {
+        MessagePassingQueue.Supplier<Object> supplier =
+            new MessagePassingQueue.Supplier<Object>() {
+                int i = 0;
+
+                @Override
+                public Object get() {
+                    return objs.get(i++);
+                }
+            };
+        int count = recvQueue.fill(supplier, objs.size());
+        metrics.notifyArrivals(count);
+        return count;
+    }
+
+    private Inserter getInserter() {
+        Inserter inserter;
+        if (producerBatchSz > 1) {
+            inserter = thdLocalBatcher.get();
+            if (inserter == null) {
+                BatchInserter b = new BatchInserter(this, producerBatchSz);
```

Comment:
Just curious: Is there specific reason to have temporary local variable `b` instead of directly using local `inserter` variable?

Reply:
To avoid casting. 
`thdLocalBatcher.set(b)` expects a BatchInserter. `inserter` is of base type.